cli/flox/src/commands/gc.rs

```diff
@@ -1,23 +1,291 @@
+                    Some(GcProgress::Scanning)
+                } else {
+                    None
+                }
+            },
+            Scanning => {
+                if GcProgress::is_deletion_line(line) {
+                    // Need to capitalize the 'd' so it matches the case of
+                    // the other log messages.
+                    let msg = Self::capitalize_deleting_store_path_line(line);
+                    Some(GcProgress::DeletingStorePaths(msg))
+                } else {
+                    None
+                }
+            },
+            DeletingStorePaths(_msg) => {
+                if GcProgress::is_deletion_line(line) {
+                    let msg = Self::capitalize_deleting_store_path_line(line);
+                    Some(GcProgress::DeletingStorePaths(msg))
+                } else if GcProgress::is_done_line(line) {
+                    Some(Done(line.to_string()))
+                } else {
+                    None
+                }
+            },
+            Done(_msg) => None,
+        }
+    }
+
+    fn capitalize_deleting_store_path_line(line: &str) -> String {
+        let (_d, rest) = line.split_at(1);
+        let mut msg = String::new();
+        msg.push('D');
+        msg.push_str(rest);
+        msg
+    }
+}
+
+/// Runs `nix store gc` and updates the spinner so that there's an indication
+/// that garbage collection hasn't stalled.
+#[instrument(skip_all, fields(progress = "Garbage collecting package data"))]
+fn run_store_gc() -> Result<()> {
+    let mut cmd = nix_base_command();
+    // The `--debug` is intentional here. You don't get any output without it.
+    cmd.args(["store", "gc", "--debug"]);
+    cmd.stderr(Stdio::piped());
+    let mut proc = cmd.spawn().context("Failed to start GC process")?;
+    let stderr = proc
+        .stderr
+        .take()
+        .ok_or(anyhow!("Failed to get stderr from GC process"))?;
+    debug!("spawned gc process");
+
+    let mut gc_progress = GcProgress::new();
+    let reader = BufReader::new(stderr);
+    let (sender, receiver) = std::sync::mpsc::channel::<GcProgress>();
+
+    // It's necessary to keep track of the parent span (the one from the `instrument`
+    // macro) so that we can associate the spans in the spinner thread with it.
+    // Otherwise, you get a spooky spinner message.
+    let parent_span = Span::current();
+
+    // Spawn separate threads for managing the spinner and reading the output
+    // of the `nix store gc` command.
+    std::thread::scope(move |s| {
+        // This thread is responsible for updating the spinner. It's more manageable
+        // to have this as a separate thread with a small inner loop instead of
+        // inlining it into the GC log reader thread.
+        let spinner_thread: std::thread::ScopedJoinHandle<
+            '_,
+            std::result::Result<(), anyhow::Error>,
+        > = s.spawn(move || {
+            let mut inner_span = Some(
```

Comment:
NB: this doesn't actually hold the span, it holds the "guard" that's returned when you enter the span. By storing that we can drop it when _we_ want to, rather than when it goes out of scope, which lets us keep the spinner going until the next update of the progress state.